### PR TITLE
Build Enhancements: added Makefile and fixed unbound variable

### PR DESCRIPTION
**Pull request by darkmuggle**
Minor enhancements to the building process to use a Makefile. Updated
the README.md to include Makefile instructions.

### CODE REVIEW

File: hack/build-go.sh [hunk 1]
```diff
@@ -30,5 +30,5 @@ fi
 
 mkdir -p ${BIN_PATH}
 
-echo "Building ${REPO}/cmd/${WHAT} (${VERSION})"
-CGO_ENABLED=0 GOOS=${GOOS} GOARCH=${GOARCH} go build ${GOFLAGS} -ldflags "${GLDFLAGS}" -o ${BIN_PATH}/${WHAT} ${REPO}/cmd/${WHAT}
+echo "Building ${REPO}/cmd/${WHAT} (${VERSION_OVERRIDE})"
+GO_ENABLED=0 GOOS=${GOOS} GOARCH=${GOARCH} go build ${GOFLAGS} -ldflags "${GLDFLAGS}" -o ${BIN_PATH}/${WHAT} ${REPO}/cmd/${WHAT}
```
Comment (yuqi-zhang) [inline hack/build-go.sh]: Was this changed intentionally...? I don't think there is a GO_ENABLED var

File: hack/build-go.sh [hunk 2]
```diff
@@ -30,5 +30,5 @@ fi
 
 mkdir -p ${BIN_PATH}
 
-echo "Building ${REPO}/cmd/${WHAT} (${VERSION})"
-CGO_ENABLED=0 GOOS=${GOOS} GOARCH=${GOARCH} go build ${GOFLAGS} -ldflags "${GLDFLAGS}" -o ${BIN_PATH}/${WHAT} ${REPO}/cmd/${WHAT}
+echo "Building ${REPO}/cmd/${WHAT} (${VERSION_OVERRIDE})"
```
Comment (yuqi-zhang) [inline hack/build-go.sh]: I made a quick PR to fix it here: https://github.com/openshift/machine-config-operator/pull/192

File: README.md [hunk 3]
```diff
@@ -1,19 +1,27 @@
...
-## machine-config-operator
-- Build: `WHAT=machine-config-operator ./hack/build-go.sh`
+## Generic building
 
-## machine-config-server
+You can either use the Makefile or the `hack/build-go.sh` script directory to build the targets. When using `hack/build-go.sh` you will need to run it via `WHAT=<TARGET> hack/build-go.sh`
+
+### machine-config-operator
+- Makefile Build: `make machine-config-operator`
+- Script Bbuild: `WHAT=machine-config-operator hack/build-go.sh`
```
Comment (yuqi-zhang) [inline README.md]: extra "b"

File: README.md [hunk 4]
```diff
@@ -1,19 +1,27 @@
...
+
+### machine-config-server
 - [Design doc](docs/MachineConfigServer.md)
-- Build: `WHAT=machine-config-server ./hack/build-go.sh`
+- Makefile Build: `make machine-config-server`
+- Script Build: `WHAT=machine-config-operator hack/build-go.sh`
```
Comment (yuqi-zhang) [inline README.md]: machine-config-server

File: README.md [hunk 5]
```diff
@@ -1,19 +1,27 @@
...
 
-## machine-config-daemon
+### machine-config-daemon
 - [Design doc](docs/MachineConfigDaemon.md)
-- Build: `WHAT=machine-config-daemon ./hack/build-go.sh`
+- Makefile Build: `make machine-config-daemon`
+- Script Build: `WHAT=machine-config-operator hack/build-go.sh`
```
Comment (yuqi-zhang) [inline README.md]: machine-config-daemon

**Comment by ashcrow** [comment]
I am personally in favor of a standard `Makefile` for building. At one point it was started in https://github.com/openshift/machine-config-operator/pull/6. Adding in @abhinavdahiya for review.

File: hack/build-go.sh (hunk 1 above)
Comment (darkmuggle) [reply hack/build-go.sh]: Fixed and re-pushed.

File: hack/build-go.sh (hunk 2 above)
Comment (darkmuggle) [reply hack/build-go.sh]: Looks like we did the same thing....

File: README.md (hunk 3 above)
Comment (darkmuggle) [reply README.md]: Fixed and repushed.

File: README.md (hunk 5 above)
Comment (darkmuggle) [reply README.md]: Fixed and repushed.

**Comment by darkmuggle** [comment]
I'll update with some of the option from @ashcrow  Makefile.

File: lib/resourceread/machineconfig.go [hunk 6]
```diff
@@ -17,13 +17,22 @@ func init() {
 	}
 }
 
+// ReadMachineConfigV1 reads MachineConfig object from bytes. Returns error.
+func ReadMachineConfigV1(objBytes []byte) (*mcfgv1.MachineConfig, error) {
```
Comment (abhinavdahiya) [inline lib/resourceread/machineconfig.go]: hmm, doesn't seem like part of this PR.

**Comment by abhinavdahiya** [comment]
I don't personally see the benefit of the Makefile. 
the targets are not correct (in Makefile sense, they are usually files on disk). Go is not great fit with Makefiles. 
But what ever the OS teams think is appropriate for the repo.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+images: image-machine-config-operator image-machine-config-server image-machine-config-daemon image-machine-config-controller
+images.rhel7: image-machine-config-operator.rhel7 image-machine-config-server.rhel7 image-machine-config-daemon.rhel7 image-machine-config-controller.rhel7
+
+clean:
+	@rm -rf $(my_p)/_output
+
+_build-%:
+	WHAT=machine-config-$* $(my_p)/hack/build-go.sh
+
+_image-%:
```
Comment (jlebon) [inline Makefile]: Hmm, those should depend on their corresponding build targets, right?

File: README.md [hunk 8]
```diff
@@ -1,19 +1,27 @@
 # machine-config-operator
 
-## machine-config-operator
-- Build: `WHAT=machine-config-operator ./hack/build-go.sh`
+## Generic building
 
-## machine-config-server
+You can either use the Makefile or the `hack/build-go.sh` script directory to build the targets. When using `hack/build-go.sh` you will need to run it via `WHAT=<TARGET> hack/build-go.sh`
```
Comment (jlebon) [inline README.md]: Hmm, not sure it's worth documenting both if the goal is to simplify things. Let's just document the `make` way?

File: lib/resourceread/machineconfig.go (hunk 6 above)
Comment (darkmuggle) [reply lib/resourceread/machineconfig.go]: Er, fixed and removed.

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,78 @@
+# vim: noexpandtab ts=8
+my_p=$(shell pwd -P)
+export GOPATH=$(shell echo $${GOPATH:-$$HOME/go})
+
+.PHONY: clean test verify update
+# Remove build artifaces
+# Example:
+#    make cllean
```
Comment (jlebon) [inline Makefile]: Minor: typo

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+verify: test machine-configs
+	$(my_p)/hack/verify-style.sh
+	$(my_p)/hack/verify-codegen.sh
+
+define target_template =
+ .PHONY: machine-config-$(1)
+ machine-config-$(1): _machine-config-$(1)
+
+ .PHONY: image-machine-config-$(1)
+ image-machine-config-$(1): _image-machine-config-$(1)
```
Comment (jlebon) [inline Makefile]: Nice use of templates!
We can add the dependency here, right? E.g.

```
  image-machine-config-$(1): _image-machine-config-$(1) _machine-config-$(1)
```

File: README.md (hunk 8 above)
Comment (jlebon) [reply README.md]: > Hmm, not sure it's worth documenting both if the goal is to simplify things. Let's just document the make way?

Any thoughts about this?

File: Makefile (hunk 10 above)
Comment (darkmuggle) [reply Makefile]: Done, take another look. I reworked it a bit to make it slightly better. Specifically, the Makefile is more dynamic in the targets: any new `Dockerfiles.X` or new `cmd/X` that gets dropped in get Makefile targets automatically.

File: Makefile (hunk 9 above)
Comment (darkmuggle) [reply Makefile]: Fixed :)

**Comment by ashcrow** [comment]
/test e2e-aws

Flake.

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+# Template for defining build targets for binaries.
+define target_template =
+ .PHONY: $(1)
+ $(1): _build-$(1)
+
+ mc += $(1)
+endef
+
+# Create a target for each command defined in `cmd` except for common.
+$(foreach P, $(patsubst cmd/%, %, $(shell find cmd -maxdepth 1 -mindepth 1 -type d -not -path cmd/common)), $(eval $(call target_template,$(P))))
```
Comment (cgwalters) [inline Makefile]: I see you are a level 50 GNU Make sorcerer too!

**Comment by darkmuggle** [comment]
/test e2e-aws

Flake.

**Comment by darkmuggle** [comment]
/test e2e-aws

Flake.

**Comment by ashcrow** [comment]
```
net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

Failure to build

/test e2e-aws

**Comment by ashcrow** [comment]
/test e2e-aws

Router didn't get created.

File: README.md (hunk 8 above)
Comment (darkmuggle) [reply README.md]: Removed, an only documented the Makefile convetion.

**Comment by ashcrow** [comment]
Should merge in a bit!

**Pull request merged**